engines/trecision/trecision.cpp

```diff
@@ -251,27 +253,37 @@ void TrecisionEngine::eventLoop() {
 			_mouseRightBtn = true;
 			break;
 
-		case Common::EVENT_KEYUP:
-			_curKey = event.kbd.keycode;
-			_curAscii = event.kbd.ascii;
-			switch (event.kbd.keycode) {
-			case Common::KEYCODE_p:
+		case Common::EVENT_CUSTOM_ENGINE_ACTION_END:
+			_curAction = event.customType;
+			switch (event.customType) {
+			case kActionFastWalk:
+				_fastWalk ^= true;
+				break;
+			case kActionPause:
 				if (!_gamePaused && !_keybInput) {
 					_curKey = Common::KEYCODE_INVALID;
+					_curAction = kActionNone;
+					keymapper->getKeymap("game-shortcuts")->setEnabled(false);
+
 					_gamePaused = true;
 					waitKey();
 				}
+
+				keymapper->getKeymap("game-shortcuts")->setEnabled(true);
 				_gamePaused = false;
 				break;
 
-			case Common::KEYCODE_CAPSLOCK:
-				_fastWalk ^= true;
-				break;
 			default:
 				break;
 			}
-			break;
+			return;
+
+		case Common::EVENT_KEYUP:
+			_curKey = event.kbd.keycode;
+			_curAscii = event.kbd.ascii;
 
+		case Common::EVENT_JOYBUTTON_UP:
+			_joyUnpause = true;
```

Comment:
I believe we don't need a new variable here, as we could do something like we do for kActionPause above:

					_curKey = Common::KEYCODE_INVALID;
					_curAction = kActionNone;

If we do need this variable, a more fitting name would be `_joyButtonUp` or `_joyButton`